docs: add ops guide for overlapping cidr ranges (#12637)
gardener/gardener: docs/operations/overlapping-network-ranges.md

control plane network ranges do not overlap. This limitation can be problematic in some cases where the network 
requirements of the shoot make a placement on existing seed clusters impossible. In the past, this has led to the costly
mitigation of deploying additional seed clusters with non-overlapping network ranges.
This document describes how Gardener handles overlapping network ranges in shoot clusters, 
which allows for more flexibility in network design and shoot placement.

## How it works
Inside the VPN tunnel, seed pod IPs and shoot IPs are disentangled by using a double-NAT mechanism, 
i.e., both the source and destination IPs are translated as shown in the following diagram:

![double-nat](./images/double-nat.png)

The NAT ranges are reserved from the 240/4 range as defined in [RFC 6890](https://datatracker.ietf.org/doc/html/rfc6890#section-2.7.4)
and thus cannot be used as shoot ranges. You can find the specific ranges in the [user networking guide](https://github.com/gardener/gardener/blob/master/docs/usage/networking/shoot_networking.md#reserved-networks)

> [!IMPORTANT]

> > The double-NAT mechanism only applies to IPv4 networks. For dual-stack or IPv6-only shoots, the shoot CIDRs must not overlap with the seed CIDRs. However, since the IPv6 range is usually assigned by the cloud provider, this is not a problem in practice.

Reviewer: This sentence sounds confusing to me. In case of dual-stack IPv4 addresses can't overlap?

Author: I've changed the section to make it clear that only IPv6 ranges must not overlap.

# Overlapping Network Ranges

## Motivation

> Shoot clusters, and their control planes are deployed in a network managed by the Gardener. The control plane can
> access shoot pods, services and nodes via a VPN connection. This makes it necessary that both the shoot and the 

Reviewer 2: nit: commas
```suggestion
Shoot clusters and their control planes are deployed in a network managed by the Gardener. The control plane can
access shoot pods, services, and nodes via a VPN connection. This makes it necessary that both the shoot and the 
```

control plane network ranges do not overlap. This limitation can be problematic in some cases where the network 
requirements of the shoot make a placement on existing seed clusters impossible. In the past, this has led to the costly
mitigation of deploying additional seed clusters with non-overlapping network ranges.
This document describes how Gardener handles overlapping network ranges in shoot clusters, 

> which allows for more flexibility in network design and shoot placement.

Reviewer 2: nit: We could simplify the sentence a bit
```suggestion
which allows for a more flexible network design and shoot placement.
```

## How it works
Inside the VPN tunnel, seed pod IPs and shoot IPs are disentangled by using a double-NAT mechanism, 
i.e., both the source and destination IPs are translated as shown in the following diagram:

![double-nat](./images/double-nat.png)

> The NAT ranges are reserved from the 240/4 range as defined in [RFC 6890](https://datatracker.ietf.org/doc/html/rfc6890#section-2.7.4)

Reviewer 2: nit: formatting
```suggestion
The NAT ranges are reserved from the `240/4` range as defined in [RFC 6890](https://datatracker.ietf.org/doc/html/rfc6890#section-2.7.4)
```

and thus cannot be used as shoot ranges. You can find the specific ranges in the [user networking guide](https://github.com/gardener/gardener/blob/master/docs/usage/networking/shoot_networking.md#reserved-networks)

> [!IMPORTANT]
> The double-NAT mechanism only applies to the IPv4 family.

> > For dual-stack or IPv6-only shoots, the IPv6 CIDRs still must not overlap with the seed CIDRs.

Reviewer 2: nit: reads better without "still"
```suggestion
> For dual-stack or IPv6-only shoots, the IPv6 CIDRs must not overlap with the seed CIDRs.
```

> However, since the IPv6 ranges are usually assigned by the cloud provider, this is not a problem in practice.

The double-NATing is achieved by IPtables rules on both ends of the VPN tunnel. The end user does not need to configure anything, as Gardener automatically applies the double-NAT mechanism when the shoot is created.

> The NAT IPs from the 240/4 range also don't show up outside the VPN tunnel, so they are not visible to the end user.

Reviewer 2: nit: formatting
```suggestion
The NAT IPs from the `240/4` range also don't show up outside the VPN tunnel, so they are not visible to the end user.
```

## HA vs. Non-HA Shoots
The double-NAT mechanism is applied to both HA and non-HA shoots. However, the implementation differs slightly:

### Non-HA Shoots

> Non-HA shoots deploy a single VPN server with an Envoy proxy in front of it. Kube-Apiserver connects to the shoot networks via this proxy.
> OpenVPN operates in client-server mode, where there is one vpn client in the shoot connecting to the vpn server in the seed.
> Kube-Apiserver is not connected directly to the VPN network.
> 
> ### HA Shoots
> [HA shoots](https://github.com/gardener/gardener/blob/master/docs/usage/high-availability/shoot_high_availability.md) deploy two VPN servers as a StatefulSet. Kube-Apiserver is connected directly to the VPN network via two VPN clients, one for each VPN server.

Reviewer 2: In other places, we usually write "Kube API server" or `kube-apiserver`. I'd probably go with the first option here:
```suggestion
Non-HA shoots deploy a single VPN server with an Envoy proxy in front of it. The Kube API server connects to the shoot networks via this proxy.
OpenVPN operates in client-server mode, where there is one vpn client in the shoot connecting to the vpn server in the seed.
The Kube API server is not connected directly to the VPN network.

### HA Shoots
[HA shoots](https://github.com/gardener/gardener/blob/master/docs/usage/high-availability/shoot_high_availability.md) deploy two VPN servers as a StatefulSet. The Kube API server is connected directly to the VPN network via two VPN clients, one for each VPN server.
```
